### ISSUE TITLE
ppo_continuous_action doesn't work on Pendulum-v1?

### ISSUE THREAD

**Issue by RishiMalhotra920** [vwxyzjn/cleanrl#504]
I've tried running CleanRL/PPO on Pendulum-v1 and I find it impossible to get an optimal solution. However, similar experiments on stable-baselines3/ppo worked. 

I've used the CleanRL/ppo command below and several hyperparameter variants. I have tried toggling the annealing, clip_vloss but I just don't see where I am going wrong.

```bash
python -m cleanrl.ppo_continuous_action --env-id Pendulum-v1 --exp-name ppo --seed 1 --track --wandb-project-name cleanRL --gamma 0.99 --vf-coef 0.5 --ent-coef 0.01 --norm-adv --num-envs
4  --clip-coef 0.2 --num-steps 2048 --clip-vloss --gae-lambda 0.95 --learning-rate 3e-4 --anneal-lr --max-grad-norm 0.5 --update-epochs 20 --num-minibatches 4 --total-timesteps 5000000 --torch-deterministic --capture-video --exp-name testing_video --target-kl 0.01
```

The script below with stable-baselines3 works achieves the optimal solution. However, sb3 is too complex to understand which is why I opted in for cleanRL.

```python
import os

import gymnasium as gym
from stable_baselines3 import PPO


def train():
    # Create environment (no render during training)
    env = gym.make("Pendulum-v1")

    # Create the PPO model with tuned hyperparameters
    model = PPO(
        "MlpPolicy",
        env,
        # Larger network
        policy_kwargs=dict(net_arch=dict(pi=[256, 256], vf=[256, 256])),
        # Learning parameters
        learning_rate=1e-3,
        n_steps=2048,
        batch_size=64,
        n_epochs=20,
        gamma=0.99,
        gae_lambda=0.95,
        clip_range=0.2,
        # Increase entropy coefficient to encourage exploration
        ent_coef=0.01,
        verbose=1,
        device="auto",
    )

    # Train for longer (200k steps)
    model.learn(total_timesteps=200000)

    # Save the model
    os.makedirs("models", exist_ok=True)
    model.save("models/ppo_pendulum")

    env.close()
```

In addition, cleanRL/rpo_continuous works too.

```
python -m cleanrl.rpo_continuous_action --no-cuda --track --wandb-project-name cleanRL --env-id Pendulum-v1 --seed 5
```

So what's the issue with cleanRL/ppo_continuous_action? Am I missing something?

**Comment by noahfarr** [comment]
Did you try running CleanRL PPO with the sb3 hyperparameters?

**Comment by RishiMalhotra920** [comment]
Yes, I tried that @noahfarr and @vwxyzjn. That would be the same command as above but with lr 1e-3 instead of 3e-4 and hidden layers of 256, 256 instead of 64, 64. 

I must be missing something critical because the rewards aren't increasing. Policy loss is around 0 and VF loss converges.

**Comment by noahfarr** [comment]
@RishiMalhotra920 did you check the NN architecture ? Are you using the same activations and weight initializations?

**Comment by RishiMalhotra920** [comment]
Yes I have. in sb3,

```python
pi = self.action_net(self.mlp_extractor.policy_net(inputs))
vf = self.value_net(self.mlp_extractor.value_net(inputs))
```

self.mlp_extractor is MlpExtractor(
  (policy_net): Sequential(
    (0): Linear(in_features=3, out_features=256, bias=True)
    (1): Tanh()
    (2): Linear(in_features=256, out_features=256, bias=True)
    (3): Tanh()
  )
  (value_net): Sequential(
    (0): Linear(in_features=3, out_features=256, bias=True)
    (1): Tanh()
    (2): Linear(in_features=256, out_features=256, bias=True)
    (3): Tanh()
  )
)
self.value_net is Linear(in_features=256, out_features=1, bias=True)
self.action_net is Linear(in_features=256, out_features=1, bias=True)
```

In cleanRL, i have:

```
        self.critic = nn.Sequential(
            layer_init(nn.Linear(np.array(envs.single_observation_space.shape).prod(), hidden_dim)),
            nn.Tanh(),
            layer_init(nn.Linear(hidden_dim, hidden_dim)),
            nn.Tanh(),
            layer_init(nn.Linear(hidden_dim, 1), std=1.0),
        )
        self.actor_mean = nn.Sequential(
            layer_init(nn.Linear(np.array(envs.single_observation_space.shape).prod(), hidden_dim)),
            nn.Tanh(),
            layer_init(nn.Linear(hidden_dim, hidden_dim)),
            nn.Tanh(),
            layer_init(nn.Linear(hidden_dim, np.prod(envs.single_action_space.shape)), std=0.01),
            nn.Tanh(),
        )
        self.actor_logstd = nn.Parameter(torch.zeros(1, np.prod(envs.single_action_space.shape)))
```

they both also use state-independent log-std for the policy net.

**Comment by noahfarr** [comment]
What about weight initialization in sb3?

**Comment by RishiMalhotra920** [comment]
They use orthogonal initialization, just like cleanRL.

**Comment by noahfarr** [comment]
Also make sure the CleanRL uses the same wrappers around the environment.

**Comment by RishiMalhotra920** [comment]
@noahfarr yeah same initialization:

CleanRL:
```python
        hidden_dim = 256
        self.critic = nn.Sequential(
            layer_init(nn.Linear(np.array(envs.single_observation_space.shape).prod(), hidden_dim)),
            nn.Tanh(),
            layer_init(nn.Linear(hidden_dim, hidden_dim)),
            nn.Tanh(),
            layer_init(nn.Linear(hidden_dim, 1), std=1.0),
        )
        self.actor_mean = nn.Sequential(
            layer_init(nn.Linear(np.array(envs.single_observation_space.shape).prod(), hidden_dim)),
            nn.Tanh(),
            layer_init(nn.Linear(hidden_dim, hidden_dim)),
            nn.Tanh(),
            layer_init(nn.Linear(hidden_dim, np.prod(envs.single_action_space.shape)), std=0.01),
        )
        self.actor_logstd = nn.Parameter(torch.zeros(1, np.prod(envs.single_action_space.shape)))
```


SB3:

```
if self.ortho_init:
            # TODO: check for features_extractor
            # Values from stable-baselines.
            # features_extractor/mlp values are
            # originally from openai/baselines (default gains/init_scales).
            module_gains = {
                self.features_extractor: np.sqrt(2),
                self.mlp_extractor: np.sqrt(2),
                self.action_net: 0.01,
                self.value_net: 1,
            }
            if not self.share_features_extractor:
                # Note(antonin): this is to keep SB3 results
                # consistent, see GH#1148
                del module_gains[self.features_extractor]
                module_gains[self.pi_features_extractor] = np.sqrt(2)
                module_gains[self.vf_features_extractor] = np.sqrt(2)

            for module, gain in module_gains.items():
                module.apply(partial(self.init_weights, gain=gain))
```

**Comment by noahfarr** [comment]
@RishiMalhotra920 and did you check whether sb3 uses the same environment wrappers as CleanRL

**Comment by RishiMalhotra920** [comment]
<img width="1156" alt="Image" src="https://github.com/user-attachments/assets/d1cadde5-bc72-425e-aeeb-92d2d83ba1e8" />

<img width="1154" alt="Image" src="https://github.com/user-attachments/assets/ad71503e-5d54-43ac-aa36-c8038e331033" />


FWIW, here's graphs from my training run with this command. This is intended to be an exact replication of the SB3 script i shared in my first post on this issue.

```bash
python -m cleanrl.ppo_continuous_action --env-id Pendulum-v1 --exp-name ppo --seed 1 --track --wandb-project-name cleanRL --gamma 0.99 --vf-coef 0.5 --ent-coef 0.01 --norm-adv --num-envs 4  --clip-coef 0.2 --num-steps 2048 --clip-vloss --gae-lambda 0.95 --learning-rate 1e-3 --no-anneal-lr --max-grad-norm 0.5 --update-epochs 20 --num-minibatches 32 --total-timesteps 5000000 --torch-deterministic --capture-video --exp-name adjusted_minibatch_size
```

**Comment by RishiMalhotra920** [comment]
Figure 3 on page 7 of this paper: https://arxiv.org/pdf/2212.07536 also shows that CleanRL PPO couldn't solve this environment. But i can verify that SB3 PPO can solve this environment.

**Comment by RishiMalhotra920** [comment]
@vwxyzjn any ideas on what im doing wrong? @noahfarr feel free to chime in!

**Comment by noahfarr** [comment]
I feel like at this point it is probably some subtle implementation difference between sb3 and CleanRL PPO like maybe different truncation handling or something like that.

**Comment by RishiMalhotra920** [comment]
Do you really think something as small as that is causing an issue? Do you think that this PPO implementation would work if i did a grid hyperparam search with optuna or something else?

**Comment by noahfarr** [comment]
According to this reply in #198  it seems like it could make quite a difference.

> Just chiming in here to agree with [this comment](https://github.com/luchris429/purejaxrl/issues/20#issuecomment-2041015222) in a related thread: I have experienced that proper timeout handling can make quite a difference. Thought I'd share my experience and solution here in case it's useful to others, and to encourage getting it implemented in CleanRL. (In relation to the discussions above I am not sure about the best way to do this though.)
> 
> Anyway: I have a vectorised environment that I was getting good results with from SB3, but to speed things up I wanted to move both environment and RL to GPU, building from CleanRL [`ppo_continuous_action_isaacgym.py`](https://github.com/vwxyzjn/cleanrl/blob/8cbca61360ef98660f149e3d76762350ce613323/cleanrl/ppo_continuous_action_isaacgym/ppo_continuous_action_isaacgym.py). However, the CleanRL PPO did not replicate the good SB3 PPO results, After much head-scratching, I discovered that disabling the timeout handling in SB3 (by no longer reporting truncations and not passing terminal observations) got SB3 stuck in the same suboptimum as CleanRL, and after adding bootstrapping at truncation to the CleanRL implementation I now get very similar good results with CleanRL to what I was getting with SB3 out of the box.
> 
> For reference, in my environment the agent can incur a lot of effort penalties early on while it's trying to learn how to get positive rewards, so early on rewards will be mostly negative. My understanding of what was happening here is that without truncation bootstrapping, just giving up and doing nothing becomes a very strong suboptimum, because every now and then the agent still gets a random (from its Markov perspective) treat of reaching the episode truncation, where without bootstrapping the reached state seems to have about zero value, which is much better than the negative values incurred when trying to learn. And the value function learning becomes noisy and slow, because there is this weird zero value that happens unpredictably every now and then. (Apols if this is obvious to others - it took me a while to suss it out.)
> 
> I based my bootstrap implementation on the SB3 one, as referenced in Costa's [comment above](https://github.com/vwxyzjn/cleanrl/issues/198#issuecomment-1153062556), changing [this line](https://github.com/vwxyzjn/cleanrl/blob/8cbca61360ef98660f149e3d76762350ce613323/cleanrl/ppo_continuous_action_isaacgym/ppo_continuous_action_isaacgym.py#L290):
> 
>             next_obs, rewards[step], next_done, info = envs.step(action)
> to:
> 
>             next_obs, rewards[step], terminations, truncations, info = envs.step(action)
>             next_done = (terminations | truncations).long()
>             with torch.no_grad():
>                 terminal_values = agent.critic(info['final_observation']).squeeze()
>             rewards[step] = torch.where(
>                 truncations, rewards[step] + args.gamma * terminal_values, rewards[step])
> This approach directly modifies the rewards, as discussed in the thread above, but that doesn't matter to me in this case.
> 
> Note that I am using my own environment here, not an IsaacGym environment, so the `info['final_observation']` is created by me in my environment class. Based on my (possibly imperfect?) interpretation of [`gymnasium.vector.VecEnv`](https://gymnasium.farama.org/api/vector/#gymnasium.vector.VectorEnv.step) I am passing this as the observation tensor (same shape as `next_obs`), but with the information from before terminated/truncated environments have been reset.

**Comment by noahfarr** [comment]
@RishiMalhotra920 I just tested ppo_continuous_action_truncted.py from #448 and it seems to work better for Pendulum-v1. Its reliably reaching around -200 to 0 reward.

**Comment by RishiMalhotra920** [comment]
This works for me. Thanks for helping Noah!
